### PR TITLE
Milestone 1 tests development

### LINKED ISSUE #4: Initialize sensors and publish their data
As a user, I should be able to read data off of all applicable sensors, so that I can see speed, distance, bearing, and the gyroscope's reading.

<br>

**Acceptance Criteria:**
- [x] All of the car's sensors - odometer, ultrasonic, infrared, gyroscope, camera, motor - are connected to the car.
- [x] Computed car stats like speed and distance traveled are published to MQTT topics
- [x] The sensors publish its data on its MQTT topic
- [x] The implementation is tested
- [x] This functionality is added to the java sdk
- [x] This integration in the java sdk is tested

### LINKED ISSUE #5: Add blinkers to the car
As a user, I should be able to use a blinker to indicate the direction that I'm turning so that I can make it a habit so that I can pass my driving test.

<br>

**Acceptance Criteria:**
- [x] Toggle the blinkers off when the angle of turning is in the opposite direction
- [x] The car listens to blinking commands via a MQTT topic (left / right / off)
- [x] The implementation is tested
- [x] This functionality is added to the java sdk
- [x] This integration in the java sdk is tested

### LINKED ISSUE #6: Add emergency stop to the car
As a user, I should be able to activate the emergency stop so that I may avoid an accident.

<br>

**Acceptance Criteria:**
- [x] Brings the car to a stop, requires that the user have to manually resume driving 
- [x] The car listens to emergency stop commands via a MQTT topic
- [x] The implementation is tested
- [x] This functionality is added to the java sdk
- [x] This integration in the java sdk is tested

### LINKED ISSUE #7: Add steering to the car
As a user, I should be able to steer so that I can change the path of my vehicle.

<br>

**Acceptance Criteria:**
- [x] Car can drive while turning
- [x] The car listens to steering commands via a MQTT topic (degrees)
- [x] The implementation is tested
- [x] This functionality is added to the java sdk
- [x] This integration in the java sdk is tested

### LINKED ISSUE #8: Add acceleration/deceleration to the car
As a user, I should be able to increase and decrease the speed of my car, so that it can go both forwards and backwards.

<br>

**Acceptance Criteria:**
- [x] Acceleration and deceleration should be possible
- [x] The car listens to acceleration/deceleration commands via a MQTT topic (signed percentages)
- [x] The implementation is tested
- [x] This functionality is added to the java sdk
- [x] This integration in the java sdk is tested

### PR DESCRIPTION
**Pull request by Lancear**
Closes #4 
Closes #5 
Closes #6 
Closes #7 
Closes #8

### CODE REVIEW

File: arduino/smartcar/smartcar.ino [hunk 1]
```diff
@@ -1,83 +1,200 @@
...
+// Sensors
+SR04 frontUltrasonic(arduinoRuntime, TRIGGER_PIN, ECHO_PIN, MAX_DISTANCE);
+GP2Y0A21 frontIR(arduinoRuntime, 0);
+GY50 gyroscope(arduinoRuntime, 0, 100UL);
+DirectionlessOdometer leftOdometer(
+    arduinoRuntime, smartcarlib::pins::v2::leftOdometerPin,
+    []() { leftOdometer.update(); }, PULSES_PER_M);
+DirectionlessOdometer rightOdometer(
+    arduinoRuntime, smartcarlib::pins::v2::rightOdometerPin,
+    []() { leftOdometer.update(); }, PULSES_PER_M);
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino]: Why aren't you creating a `SmartCar` but instead using a `SimpleCar` with sensors on the side?

File: arduino/smartcar/smartcar.ino [hunk 2]
```diff
@@ -1,83 +1,200 @@
...
+      // Publishing average distance travelled via both Odometers
+      //  (distance logged is in centimeters)
+      const auto avg_distance = String(
+          (leftOdometer.getDistance() + rightOdometer.getDistance()) / 2);
+      mqtt.publish(ODOMETER_DISTANCE, avg_distance);
+
+      // Publishing average speed of the car via both Odometers
+      //  (speed logged is in meters per second)
+      const auto avg_speed =
+          String((leftOdometer.getSpeed() + rightOdometer.getSpeed()) / 2);
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino]: If you have a `SmartCar` you can instead do `car.getSpeed()`)

File: arduino/smartcar/smartcar.ino [hunk 3]
```diff
@@ -1,83 +1,200 @@
...
+      mqtt.publish(ODOMETER_SPEED, avg_speed);
+
+      // Publish heading angle (degrees: [0, 360]) using gyroscope
+      const auto currentHeading = gyroscope.getHeading();
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino]: Similarly here `car.getHeading()` if the `car` is a `SmartCar` instance

**Comment by Lancear** [comment]
image
🎉 🙌 🎉

File: arduino/smartcar/smartcar.ino (hunk 2 above)
Comment (Lancear) [reply arduino/smartcar/smartcar.ino]: Thanks, we will do!
btw, is there a way to get the throttles percentage? We couldn't find anything in the docs for BrushedMotor.

Comment (platisd) [reply arduino/smartcar/smartcar.ino]: No, if you want to keep track of the speed you have set you can... keep track of it yourself. :stuck_out_tongue:

**Review by plepisnew — DISMISSED** [review]

**Review by GeorgZs — DISMISSED** [review]

**Review by IVidackovic — APPROVED** [review]

**Review by bardiaForooraghi — APPROVED** [review]

**Review by plepisnew — APPROVED** [review]

**Pull request merged**